include/potter/potter.hpp

```diff
@@ -521,6 +639,52 @@ class Integrator {
             }
             break;
         }
+		case 4:
+		{
+			// Fourth virial coefficient: three parts B_4_1,B_4_2,B_4_3
+			// The integrand functions
+			//typedef int (*integrand) (unsigned ndim, const double *x, void *, unsigned fdim, double* fval);
+			auto cubature_integrand_1 = [](unsigned ndim, const double* x, void* p_shared_data, unsigned fdim, double* fval) {
+				auto& shared = *((class SharedDataBase<double, TEMPTYPE>*)(p_shared_data));
+				shared.atomic_B4_1_integrand(x[0], x[1], x[2], x[3], x[4], fval);
+				return 0; // success
+			};
+
+			auto cubature_integrand_2 = [](unsigned ndim, const double* x, void* p_shared_data, unsigned fdim, double* fval) {
+				auto& shared = *((class SharedDataBase<double, TEMPTYPE>*)(p_shared_data));
+				shared.atomic_B4_2_integrand(x[0], x[1], x[2], x[3], x[4], fval);
+				return 0; // success
+			};
+
+			auto cubature_integrand_3 = [](unsigned ndim, const double* x, void* p_shared_data, unsigned fdim, double* fval) {
+				auto& shared = *((class SharedDataBase<double, TEMPTYPE>*)(p_shared_data));
+				shared.atomic_B4_3_integrand(x[0], x[1], x[2], x[3], x[4], x[5], fval);
+				return 0; // success
+			};
+
+
+			// prefactors for each contribution 
+			std::valarray<double> pre_factors = { -3.0*(27.0 / 4.0) , 3.0*(27.0 / 2.0) ,  -27.0 / (8.0*M_PI) };
+
+			auto naxes = 5; // How many dimensions the integral is taken over (r14, r13, gama, r12, eta)
+			hcubature(ndim, cubature_integrand_1, &shared, naxes, &(xmin_1[0]), &(xmax_1[0]), 1E8, 0, 1e-4, ERROR_INDIVIDUAL, &(val[0]), &(err[0]));
+			hcubature(ndim, cubature_integrand_2, &shared, naxes, &(xmin_2[0]), &(xmax_2[0]), 1E8, 0, 1e-4, ERROR_INDIVIDUAL, &(val[1]), &(err[1]));
+
+			naxes = 6;
+			hcubature(ndim, cubature_integrand_3, &shared, naxes, &(xmin_3[0]), &(xmax_3[0]), 1E8, 0, 1e-4, ERROR_INDIVIDUAL, &(val[2]), &(err[2]));
+
+
+			for (auto i = 0; i < val.size(); ++i) {
+				val[i] = pre_factors[i] * val[i];
+				err[i] = pre_factors[i] * err[i];
+			}
+
+			val[0] = val[0] + val[1] + val[2];
```

Comment:
I think this is not the way to return the val, should it be at the second position in the vector?

Reply:
Nope, the val array is used for storing more output parts (the real and imaginary parts for complex numbers, for instance).  Please see the B_2 case as a demonstration.  I think you want to have different ``std::valarray`` for each integrand, then return the sum of the arrays.  In C++, if ``v1``, ``v2`` are ``std::valarray``, you can do ``v1``+``v2``, which will be coefficient-wise sum.  That's why I used them instead of ``std::vector``.